debezium-server-voyager/debezium-server-voyagerexporter/src/main/java/io/debezium/server/ybexporter/EventQueue.java

```diff
@@ -209,28 +234,128 @@ public void close() {
     public void sync() {
         try {
             currentQueueSegment.sync();
-        }
-        catch (IOException | SQLException e) {
+        } catch (IOException | SQLException e) {
             throw new RuntimeException(e);
         }
     }
+
+    class EventDedupCache {
+        private static final String EOF_MARKER = "\\.";
+        String dataDir;
+        private LinkedList<String> mostRecentIdFirstList = new LinkedList<>();
+        private Set<String> cache = new HashSet<>();
+        private long currentQueueSegmentIndex = 0;
+        private static final String QUEUE_SEGMENT_FILE_NAME = "segment";
+        private static final String QUEUE_SEGMENT_FILE_EXTENSION = "ndjson";
+        private static final String QUEUE_FILE_DIR = "queue";
+        private long maxCacheSize = 1000000;
+        private String currentQueueSegmentPath;
+        private ObjectMapper mapper = new ObjectMapper();
+
+        public EventDedupCache(String dataDir) {
+            this.dataDir = dataDir;
+        }
+
+        public void warmUp(Map<Long, Long> totalEventsPerSegment) {
+            if (totalEventsPerSegment.size() == 0) {
+                LOGGER.info("No queue segments found. Nothing to recover into event dedup cache");
+                return;
+            }
+            cache = new HashSet<>();
+            long eventsToBeProcessed = 0;
+            // Find which segement to start reading from, the Map entries are in descending
+            // order of segment number
+            for (Map.Entry<Long, Long> entry : totalEventsPerSegment.entrySet()) {
+                currentQueueSegmentIndex = entry.getKey();
+                eventsToBeProcessed += entry.getValue();
+                if (eventsToBeProcessed >= maxCacheSize) {
+                    // We have enough events to fill the cache. Break out of loop
+                    break;
+                }
+            }
+            while (currentQueueSegmentIndex <= totalEventsPerSegment.size() - 1) {
+                currentQueueSegmentPath = getFilePathOfCurrentQueueSegment();
+                String line;
+                BufferedReader input;
+                try {
+                    input = new BufferedReader(new FileReader(currentQueueSegmentPath));
+                    while ((line = input.readLine()) != null) {
```

Comment:
Something for later: 
Reading the actual queueSegment file should ideally be the responsibility of the `QueueSegment` class. 

The EventDedupCache should probably just call something like `queueSegment.getNextEvent()`